### ISSUE TITLE
User signed-in through a successful sign-up isn't authorized to write to database.

### ISSUE THREAD

**Issue by ggaabe** [firebase/firebase-js-sdk#255]
### Environment

  * Operating System version: macOS High Sierra 10.13 Beta
  * Firebase SDK version: 4.5.2
  * Firebase Product: auth, database

### Issue:

I have a sign-up form with a success callback that writes other sign-up info to that user's node in the real-time database. Whenever the callback gets called, permission is denied. The database rules are:
```
{
  "rules": {
    ".read": "auth != null",
    ".write": "auth != null"
  }
}
```
Eventually, I moved the database write to the onAuthStateChanged callback.
Here's what I found:

1. The onAuthStateChanged callback gets called after the signup / onCreateUserWithEmail success callback.
2. If the database write is called from the onAuthStateChanged on a sign-in, it writes successfully.
3. If the database write is called from the onAuthStateChanged on a sign-in triggered by a sign up, the write attempt throws a permission denied error.

#### Relevant Code:

```
var database = firebase.database();
var signUpMode = false;
//this function writes the user info to the realtime database.
    function logUser(user, signUpPhoneNumber, department){
      database.ref('users/' + user.uid).set({
       phone: signUpPhoneNumber,
       department: department
     }).catch(function(error){
       console.log(error);
     })
    }

function handleSignUp() {
      var email = document.getElementById('emailSignUp').value;
      var password = document.getElementById('passwordSignUp').value;
      var signUpPhoneNumber = document.getElementById('phoneNumber').value;
      var department = document.getElementById('department').value;
      if (email.length < 4) {
        alert('Please enter an email address.');
        return;
      }
      if (password.length < 4) {
        alert('Please enter a password.');
        return;
      }
      if(signUpPhoneNumber.length < 7){
        alert("Please enter a valid phone number");
        return;
      }
      if(department.length < 2){
        alert("Please enter a department name or code.");
        return;
      }
      firebase.auth().createUserWithEmailAndPassword(email, password).then(function(user){
        console.log("Email: " +firebase.auth().currentUser.email); //this successfully prints the email
        signUpMode = true;
        return user;
      }).then(function(user){
        console.log("Logged in: " + firebase.auth().currentUser.email);  //this successfully prints the email as well
      }).catch(function(error) {
        var errorCode = error.code;
        var errorMessage = error.message;
        if (errorCode == 'auth/weak-password') {
          alert('The password is too weak.');
        } else {
          alert(errorMessage);
        }
        console.log(error);
      });
    }

//this function is inside of an initApp function that initializes firebase:
firebase.auth().onAuthStateChanged(function(user) {
        document.getElementById('quickstart-verify-email').disabled = true;
        if (user) {
          // User is signed in.
          console.log("Auth state changed");
          //all of the below data is not null.
          var displayName = user.displayName;
          var email = user.email;
          var emailVerified = user.emailVerified;
          var photoURL = user.photoURL;
          var isAnonymous = user.isAnonymous;
          var uid = user.uid;
          var providerData = user.providerData;
          if(signUpMode){
            var signUpPhoneNumber = document.getElementById('phoneNumber').value;
            var department = document.getElementById('department').value;
           // this fails with a permission denied error if triggered by a signup. It runs without permission denied errors when not signUpMode.
            logUser(user, signUpPhoneNumber, department);
          }

          document.getElementById('quickstart-sign-in-status').textContent = 'Signed in';
          document.getElementById('quickstart-sign-in').textContent = 'Sign out';
       
          if (!emailVerified) {
            document.getElementById('quickstart-verify-email').disabled = false;
          }
        } else {
          // User is signed out.
          document.getElementById('quickstart-sign-in-status').textContent = 'Signed out';
          document.getElementById('quickstart-sign-in').textContent = 'Sign in';
        }
        document.getElementById('quickstart-sign-in').disabled = false;
      });

```

**Comment by bojeil-google** [comment]
Sign in and sign up are the same form Auth perspective. In both cases we issue an ID token.

**Comment by jshcrowthe** [comment]
We had some issues with version `4.5.2` where some of the auth permissions weren't properly passed along. They have been fixed in version `4.6.0`. Please update and try to reproduce the issue there. If you can then feel free to re-open.
